datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java

```diff
@@ -30,136 +21,203 @@ void beforeGetConnection(CON connectionKey, String dataSourceName) {
         Span connectionSpan = tracer.nextSpan().name("jdbc:/" + dataSourceName + SleuthListenerAutoConfiguration.SPAN_CONNECTION_POSTFIX);
         connectionSpan.kind(Kind.CLIENT);
         connectionSpan.start();
-        connectionSpans.put(connectionKey, connectionSpan);
+        ConnectionInfo connectionInfo = new ConnectionInfo(new SpanWithScope(connectionSpan, tracer.withSpanInScope(connectionSpan)));
+        openConnections.put(connectionKey, connectionInfo);
     }
 
     void afterGetConnection(CON connectionKey, Throwable t) {
-        Span connectionSpan = connectionSpans.get(connectionKey);
         if (t != null) {
-            connectionSpans.remove(connectionKey);
-            connectionSpan.kind(Kind.CLIENT);
-            connectionSpan.finish();
+            ConnectionInfo connectionInfo = openConnections.remove(connectionKey);
+            connectionInfo.getSpan().finish();
```

Comment:
This one is pretty minor, do you think it should propagate error here?
```
        if (t != null) {
            connectionSpan.getSpan().error(t);
        }
```

Reply:
Good catch! Added